File: client/client-multi/client-objective-c/KaaTests/channel/DefaultOperationTcpChannelTest.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#define HC_SHORTHAND
#import <OCHamcrest/OCHamcrest.h>

#define MOCKITO_SHORTHAND
#import <OCMockito/OCMockito.h>

#import <XCTest/XCTest.h>
#import "DefaultOperationTcpChannel.h"
#import "KAASocket.h"
#import "KaaClientState.h"
#import "KeyUtils.h"
#import "AvroBytesConverter.h"
#import "GenericTransportInfo.h"
#import "TransportProtocolIdHolder.h"
#import "KAATcpSyncResponse.h"
#import "IPTransportInfo.h"
#import "KAATcpPingResponse.h"
#import "KAATcpDisconnect.h"
#import "TestsHelper.h"

#pragma mark - MockedOperationTcpChannel

@interface MockedOperationTcpChannel : DefaultOperationTcpChannel

@property (nonatomic, strong) KAASocket *socketMock;
@property (nonatomic, strong) NSInputStream *inputStream;
@property (nonatomic, strong) NSOutputStream *outputStream;

@end

@implementation MockedOperationTcpChannel

- (instancetype)initWithClientState:(id<KaaClientState>)state failoverManager:(id<FailoverManager>)failoverMgr {
    self = [super initWithClientState:state failoverManager:failoverMgr failureDelegate:nil];
    if (self) {
        CFReadStreamRef readStream = NULL;
        CFWriteStreamRef writeStream = NULL;
        CFStreamCreateBoundPair(NULL, &readStream, &writeStream, 4096);
        
        self.inputStream = (__bridge_transfer NSInputStream *)readStream;
        self.outputStream = (__bridge_transfer NSOutputStream *)writeStream;
        
        self.socketMock = mock([KAASocket class]);
        
        [given([self.socketMock input]) willReturn:self.inputStream];
        [given([self.socketMock output]) willReturn:self.outputStream];
        
        [self.inputStream open];
        [self.outputStream open];
    }
    return self;
}

- (KAASocket *)createSocket {
    return self.socketMock;
}

@end

#pragma mark - DefaultOperationTcpChannelTest

@interface DefaultOperationTcpChannelTest : XCTestCase

@end

@implementation DefaultOperationTcpChannelTest


- (void)testDefaultOperationTcpChannel {
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    id<FailoverManager> failoverManager = mockProtocol(@protocol(FailoverManager));
    id<KaaDataChannel> tcpchannel = [[DefaultOperationTcpChannel alloc] initWithClientState:state
                                                                            failoverManager:failoverManager
                                                                            failureDelegate:nil];
    XCTAssertNotNil([tcpchannel getId]);
    XCTAssertNotNil([tcpchannel getSupportedTransportTypes]);
    XCTAssertNotEqual(0, [[tcpchannel getSupportedTransportTypes] count]);
}

/**
 * The issue behind this test is that stream delegate methods don't get triggered after mocking KAASocket
 */
- (void)DISABLED_testSync {
    KeyPair *clientKeys = [KeyUtils generateKeyPair];
    id<KaaClientState> clientState = mockProtocol(@protocol(KaaClientState));
    [given([clientState privateKey]) willReturnStruct:[clientKeys getPrivateKeyRef] objCType:@encode(SecKeyRef)];
    [given([clientState publicKey]) willReturnStruct:[clientKeys getPublicKeyRef] objCType:@encode(SecKeyRef)];
    
    id<FailoverManager> failoverManager = mockProtocol(@protocol(FailoverManager));
    MockedOperationTcpChannel *tcpChannel = [[MockedOperationTcpChannel alloc] initWithClientState:clientState failoverManager:failoverManager];
    
    AvroBytesConverter *requestCreator = [[AvroBytesConverter alloc] init];
    id<KaaDataMultiplexer> multiplexer = mockProtocol(@protocol(KaaDataMultiplexer));
    [given([multiplexer compileRequestForTypes:anything()])willReturn:[requestCreator toBytes:[[SyncRequest alloc] init]]];
    id<KaaDataDemultiplexer> demultiplexer = mockProtocol(@protocol(KaaDataDemultiplexer));
    
    [tcpChannel setMultiplexer:multiplexer];
    [tcpChannel setDemultiplexer:demultiplexer];
    [tcpChannel syncForTransportType:TRANSPORT_TYPE_USER];    // will cause call to KaaDataMultiplexer.compileRequest(...) after "CONNECT" messsage
    [tcpChannel syncForTransportType:TRANSPORT_TYPE_PROFILE];
    
    [KeyUtils generateKeyPair];
    id<TransportConnectionInfo> server = [self createTestServerInfoWithServerType:SERVER_OPERATIONS transportProtocolId:[TransportProtocolIdHolder TCPTransportID] host:@"localhost" port:9009 publicKey:[KeyUtils getPublicKey]];
    
    [tcpChannel setServer:server withKeyPair:clientKeys];
    uint8_t rawConnackChar[] = {0x20, 0x02, 0x00, 0x01};
    [tcpChannel.outputStream write:rawConnackChar maxLength:sizeof(rawConnackChar)];
    
    SyncResponse *response = [[SyncResponse alloc] init];
    [response setStatus:SYNC_RESPONSE_RESULT_TYPE_SUCCESS];
    NSData *kaatcpsyncrespData = [self getNewKAATcpSyncResponseWithResponse:response];
    [tcpChannel.outputStream write:[kaatcpsyncrespData bytes] maxLength:[kaatcpsyncrespData length]];
    
    [NSThread sleepForTimeInterval:1]; // sleep a bit to let the message to be received
    [tcpChannel syncForTransportType:TRANSPORT_TYPE_USER]; // causes call to KaaDataMultiplexer.compileRequest(...) for "KAA_SYNC" messsage
    [verifyCount(multiplexer, times(2)) compileRequestForTypes:anything()];
    
    [tcpChannel syncForTransportType:TRANSPORT_TYPE_EVENT];
    [verifyCount(multiplexer, times(3)) compileRequestForTypes:anything()];
    [verifyCount(tcpChannel.socketMock, times(3)) output];
    
    [tcpChannel.outputStream write:[[[[KAATcpPingResponse alloc] init] getFrame] bytes] maxLength:[[[[KAATcpPingResponse alloc] init] getFrame] length]];
    
    [tcpChannel syncAll];
    [verifyCount(multiplexer, times(2)) compileRequestForTypes:[tcpChannel getSupportedTransportTypes]];
    
    KAATcpDisconnect *disconnect = [[KAATcpDisconnect alloc] initWithDisconnectReason:DisconnectReasonInternalError];
    [tcpChannel.outputStream write:[[disconnect getFrame] bytes] maxLength:[[disconnect getFrame] length]];
    
    [tcpChannel syncAll];
    [verifyCount(multiplexer, times(3)) compileRequestForTypes:[tcpChannel getSupportedTransportTypes]];
    [tcpChannel shutdown];
}

- (void)testConnectivity {
    KeyPair *clientKeys = [KeyUtils generateKeyPair];
    id<KaaClientState> clientState = mockProtocol(@protocol(KaaClientState));
    [given([clientState privateKey]) willReturnStruct:[clientKeys getPrivateKeyRef] objCType:@encode(SecKeyRef)];
    [given([clientState publicKey]) willReturnStruct:[clientKeys getPublicKeyRef] objCType:@encode(SecKeyRef)];
    
    id<FailoverManager> failoverManager = mockProtocol(@protocol(FailoverManager));
    DefaultOperationTcpChannel *channel = [[DefaultOperationTcpChannel alloc] initWithClientState:clientState
                                                                                  failoverManager:failoverManager
                                                                                  failureDelegate:nil];
    
    id<TransportConnectionInfo> server = [self createTestServerInfoWithServerType:SERVER_OPERATIONS transportProtocolId:[TransportProtocolIdHolder TCPTransportID] host:@"www.test.fake" port:999 publicKey:[KeyUtils getPublicKey]];
    XCTAssertNotNil(server);
    
    ConnectivityChecker *checker = mock([ConnectivityChecker class]);
    [given([checker isConnected]) willReturnBool:NO];
    [channel setConnectivityChecker:checker];
}

#pragma mark - Supporting methods

- (id<TransportConnectionInfo>)createTestServerInfoWithServerType:(ServerType)serverType
                                              transportProtocolId:(TransportProtocolId *)TPid
                                                             host:(NSString *)host
                                                             port:(uint32_t)port
                                                        publicKey:(NSData *)publicKey {
    ProtocolMetaData *md = [TestsHelper buildMetaDataWithTransportProtocolId:TPid host:host port:port publicKey:publicKey];
    return  [[GenericTransportInfo alloc] initWithServerType:serverType meta:md];
}

- (NSData *)getNewKAATcpSyncResponseWithResponse:(SyncResponse *)syncResponse {
    AvroBytesConverter *responseCreator = [[AvroBytesConverter alloc] init];
    NSData *data = [responseCreator toBytes:syncResponse];
    KAATcpSyncResponse *response = [[KAATcpSyncResponse alloc] initWithAvro:data zipped:NO encypted:NO];
    return [response getFrame];
}

@end

File: client/client-multi/client-java-core/src/test/java/org/kaaproject/kaa/client/channel/DefaultOperationTcpChannelTest.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.channel;

import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNotNull;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PipedInputStream;
import java.io.PipedOutputStream;
import java.net.Socket;
import java.net.UnknownHostException;
import java.security.KeyPair;
import java.security.NoSuchAlgorithmException;

import org.junit.Test;
import org.kaaproject.kaa.client.channel.connectivity.ConnectivityChecker;
import org.kaaproject.kaa.client.channel.failover.FailoverManager;
import org.kaaproject.kaa.client.channel.impl.channels.DefaultOperationTcpChannel;
import org.kaaproject.kaa.client.persistence.KaaClientState;
import org.kaaproject.kaa.common.TransportType;
import org.kaaproject.kaa.common.avro.AvroByteArrayConverter;
import org.kaaproject.kaa.common.channels.protocols.kaatcp.messages.Disconnect;
import org.kaaproject.kaa.common.channels.protocols.kaatcp.messages.Disconnect.DisconnectReason;
import org.kaaproject.kaa.common.channels.protocols.kaatcp.messages.PingResponse;
import org.kaaproject.kaa.common.endpoint.gen.SyncRequest;
import org.kaaproject.kaa.common.endpoint.gen.SyncResponse;
import org.kaaproject.kaa.common.endpoint.gen.SyncResponseResultType;
import org.kaaproject.kaa.common.endpoint.security.KeyUtil;
import org.mockito.Mockito;

public class DefaultOperationTcpChannelTest {

    private final KeyPair clientKeys;

    public DefaultOperationTcpChannelTest() throws Exception {
        clientKeys = KeyUtil.generateKeyPair();
    }

    class TestOperationTcpChannel extends DefaultOperationTcpChannel {
        public Socket socketMock;
        public OutputStream os;
        public InputStream is;

        public TestOperationTcpChannel(KaaClientState state,
                FailoverManager failoverManager) throws IOException {
            super(state, failoverManager, null);
            PipedInputStream in = new PipedInputStream(4096);
            PipedOutputStream out = new PipedOutputStream(in);
            os = out;
            is = in;
            socketMock = Mockito.mock(Socket.class);
            Mockito.when(socketMock.getOutputStream()).thenReturn(os);
            Mockito.when(socketMock.getInputStream()).thenReturn(is);
        }

        @Override
        protected Socket createSocket(String host, int port)
                throws UnknownHostException, IOException {
            return socketMock;
        }
        @Override
        public void shutdown() {
            super.shutdown();
        }
    }

    @Test
    public void testDefaultOperationTcpChannel() {
        KaaClientState clientState = Mockito.mock(KaaClientState.class);
        FailoverManager failoverManager = Mockito.mock(FailoverManager.class);
        KaaDataChannel tcpChannel = new DefaultOperationTcpChannel(clientState, failoverManager, null);
        assertNotNull("New channel's id is null", tcpChannel.getId());
        assertNotNull("New channel does not support any of transport types", tcpChannel.getSupportedTransportTypes());
        assertNotEquals(0, tcpChannel.getSupportedTransportTypes().size());
    }

    @Test
    public void testSync() throws Exception {
        KaaClientState clientState = Mockito.mock(KaaClientState.class);
        Mockito.when(clientState.getPrivateKey()).thenReturn(clientKeys.getPrivate());
        Mockito.when(clientState.getPublicKey()).thenReturn(clientKeys.getPublic());

        FailoverManager failoverManager = Mockito.mock(FailoverManager.class);
        TestOperationTcpChannel tcpChannel = new TestOperationTcpChannel(clientState, failoverManager);

        AvroByteArrayConverter<SyncResponse> responseCreator = new AvroByteArrayConverter<SyncResponse>(SyncResponse.class);
        AvroByteArrayConverter<SyncRequest> requestCreator = new AvroByteArrayConverter<SyncRequest>(SyncRequest.class);
        KaaDataMultiplexer multiplexer = Mockito.mock(KaaDataMultiplexer.class);
        Mockito.when(multiplexer.compileRequest(Mockito.anyMapOf(TransportType.class, ChannelDirection.class))).thenReturn(requestCreator.toByteArray(new SyncRequest()));
        KaaDataDemultiplexer demultiplexer = Mockito.mock(KaaDataDemultiplexer.class);
        tcpChannel.setMultiplexer(multiplexer);
        tcpChannel.setDemultiplexer(demultiplexer);
        tcpChannel.sync(TransportType.USER);        // will cause call to KaaDataMultiplexer.compileRequest(...) after "CONNECT" messsage
        tcpChannel.sync(TransportType.PROFILE);

        TransportConnectionInfo server = IPTransportInfoTest.createTestServerInfo(ServerType.OPERATIONS, TransportProtocolIdConstants.TCP_TRANSPORT_ID,
                "localhost", 9009, KeyUtil.generateKeyPair().getPublic());

        tcpChannel.setServer(server); // causes call to KaaDataMultiplexer.compileRequest(...) for "CONNECT" messsage
        byte [] rawConnack = new byte[] { 0x20, 0x02, 0x00, 0x01 };
        tcpChannel.os.write(rawConnack);

        SyncResponse response = new SyncResponse();
        response.setStatus(SyncResponseResultType.SUCCESS);
        tcpChannel.os.write(new org.kaaproject.kaa.common.channels.protocols.kaatcp.messages.SyncResponse(responseCreator.toByteArray(response), false, false).getFrame().array());
        Thread.sleep(1000);  // sleep a bit to let the message to be received
        tcpChannel.sync(TransportType.USER); // causes call to KaaDataMultiplexer.compileRequest(...) for "KAA_SYNC" messsage

        Mockito.verify(multiplexer, Mockito.times(2)).compileRequest(Mockito.anyMapOf(TransportType.class, ChannelDirection.class));

        tcpChannel.sync(TransportType.EVENT);

        Mockito.verify(multiplexer, Mockito.times(3)).compileRequest(Mockito.anyMapOf(TransportType.class, ChannelDirection.class));
        Mockito.verify(tcpChannel.socketMock, Mockito.times(3)).getOutputStream();

        Mockito.reset(multiplexer);
        tcpChannel.os.write(new PingResponse().getFrame().array());

        tcpChannel.syncAll();
        Mockito.verify(multiplexer, Mockito.times(1)).compileRequest(tcpChannel.getSupportedTransportTypes());
        tcpChannel.os.write(new Disconnect(DisconnectReason.INTERNAL_ERROR).getFrame().array());

        tcpChannel.syncAll();
        Mockito.verify(multiplexer, Mockito.times(1)).compileRequest(tcpChannel.getSupportedTransportTypes());
        tcpChannel.shutdown();
    }

    @Test
    public void testConnectivity() throws NoSuchAlgorithmException {
        KaaClientState clientState = Mockito.mock(KaaClientState.class);
        Mockito.when(clientState.getPrivateKey()).thenReturn(clientKeys.getPrivate());
        Mockito.when(clientState.getPublicKey()).thenReturn(clientKeys.getPublic());

        FailoverManager failoverManager = Mockito.mock(FailoverManager.class);
        DefaultOperationTcpChannel channel = new DefaultOperationTcpChannel(clientState, failoverManager, null);

        TransportConnectionInfo server = IPTransportInfoTest.createTestServerInfo(ServerType.OPERATIONS, TransportProtocolIdConstants.TCP_TRANSPORT_ID,
                "www.test.fake", 999, KeyUtil.generateKeyPair().getPublic());
        
        ConnectivityChecker checker = Mockito.mock(ConnectivityChecker.class);
        Mockito.when(checker.checkConnectivity()).thenReturn(false);

        channel.setConnectivityChecker(checker);
    }
}
